File: android_samples_java/22.5 GeofenceDemo/app/src/main/java/com/msnegi/geofencedemo/HomeFragment.java
package com.msnegi.geofencedemo;

import static android.app.Activity.RESULT_CANCELED;
import static android.app.Activity.RESULT_OK;

import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.IntentSender;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.location.Location;
import android.location.LocationManager;
import android.os.Build;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import android.speech.tts.TextToSpeech;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.PendingResult;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.common.api.Status;
import com.google.android.gms.location.Geofence;
import com.google.android.gms.location.GeofencingRequest;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.LocationSettingsRequest;
import com.google.android.gms.location.LocationSettingsResult;
import com.google.android.gms.location.LocationSettingsStatusCodes;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.CircleOptions;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.msnegi.geofencedemo.firstupdate.LatLongReceived;
import com.msnegi.geofencedemo.firstupdate.LocManager;
import com.msnegi.geofencedemo.regularupdate.LocData;

import java.util.ArrayList;
import java.util.List;

public class HomeFragment extends Fragment implements LatLongReceived, OnMapReadyCallback,
        GoogleMap.OnMapClickListener,
        GoogleMap.OnMarkerClickListener,
        GoogleMap.OnInfoWindowClickListener
{
    private static final String TAG = "GeoFence";
    GoogleApiClient googleApiClient;
    private MapView mapView;
    private GoogleMap googleMap;
    private Marker locationMarker;
    float zoomLevel = 15f;
    LocManager locationManager;

    public static final String NEW_GEOFENCE_NUMBER = BuildConfig.APPLICATION_ID + ".NEW_GEOFENCE_NUMBER";
    public static final long GEOFENCE_EXPIRATION_IN_HOURS = 365 * 24;
    public static final long GEOFENCE_EXPIRATION_IN_MILLISECONDS = GEOFENCE_EXPIRATION_IN_HOURS * 60 * 1000;
    public static final float GEOFENCE_RADIUS_IN_METERS = 100; // 100 m
    private static final int PERMISSIONS_REQUEST = 105;
    private static final int PERMISSION_REQUEST_CODE = 110;
    protected ArrayList<Geofence> mGeofenceList;
    private PendingIntent mGeofencePendingIntent;
    private SharedPreferences mSharedPreferences;
    public static final String SHARED_PREFERENCES_NAME = BuildConfig.APPLICATION_ID + ".SHARED_PREFERENCES_NAME";

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        mSharedPreferences = getActivity().getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE);
        mGeofenceList = new ArrayList<>();
        mGeofencePendingIntent = null;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        googleApiClient = new GoogleApiClient.Builder(getActivity())
                                             .addApi(LocationServices.API).build();
        googleApiClient.connect();
        locationManager = LocManager.getInstance();
        locationManager.connectGoogleClient(getActivity(), this);

        View view = inflater.inflate(R.layout.fragment_home, container, false);
        mapView = view.findViewById(R.id.map);
        mapView.onCreate(savedInstanceState);
        mapView.getMapAsync(this);
        return view;
    }

    @Override
    public void onMapReady(GoogleMap mMap)
    {
        googleMap = mMap;
        mapView.onResume();

        googleMap.setTrafficEnabled(true);
        googleMap.setOnMapClickListener(this);
        googleMap.setOnMarkerClickListener(this);
        googleMap.setOnInfoWindowClickListener(this);
    }

    @Override
    public void onLocationReceived(Location loc) {
        System.out.println("first Lat : " + loc.getLatitude() + " Long : " + loc.getLongitude());
        setupMap(loc.getLatitude(), loc.getLongitude());
    }

    private void setupMap(double mLat, double mLong) {
        LatLng mLoc = new LatLng(mLat, mLong);

        googleMap.clear();
        MarkerOptions options = new MarkerOptions();
        options.position(mLoc);
        options.icon(BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_RED));
        googleMap.addMarker(options);
        googleMap.getUiSettings().setZoomControlsEnabled(true);
        googleMap.moveCamera(CameraUpdateFactory.newLatLng(mLoc));
        googleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(new LatLng(mLat, mLong), 5));

        CameraPosition cameraPosition = new CameraPosition.Builder()
                .target(new LatLng(mLat, mLong))
                .zoom(zoomLevel)
                .build();
        googleMap.animateCamera(CameraUpdateFactory.newCameraPosition(cameraPosition));
        mapView.onResume();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mapView.onDestroy();
    }

    //----------------- Geofence setup ----------------

    private int getNewGeofenceNumber() {
        int number = mSharedPreferences.getInt(NEW_GEOFENCE_NUMBER, 0);
        SharedPreferences.Editor editor = mSharedPreferences.edit();
        editor.putInt(NEW_GEOFENCE_NUMBER, number + 1);
        editor.commit();
        return number;
    }

    @Override
    public void onMapClick(LatLng latLng) {
        final String key = getNewGeofenceNumber() + "";
        createGeofence(new LatLng(latLng.latitude, latLng.longitude),latLng.latitude,latLng.longitude,100,365*24*60*60*1000,key);
    }

    public void createGeofence(LatLng latLng, final double lat, final double lng, final int radius, final int expire, final String key)
    {
        //-- draw a circle on map.
        googleMap.addMarker(new MarkerOptions()
                .title("G:" + key)
                .snippet("Click here if you want delete this geofence")
                .position(latLng));

        googleMap.addCircle(new CircleOptions()
                .center(latLng)
                .radius(GEOFENCE_RADIUS_IN_METERS)
                .strokeColor(Color.RED)
                .fillColor(Color.parseColor("#80ff0000")));

        //final String key = getNewGeofenceNumber() + "";
        final long expTime = System.currentTimeMillis() + GEOFENCE_EXPIRATION_IN_MILLISECONDS;

        Geofence geofence = new Geofence.Builder()
                .setRequestId(key)
                .setCircularRegion(
                        latLng.latitude,
                        latLng.longitude,
                        GEOFENCE_RADIUS_IN_METERS)
                .setExpirationDuration(GEOFENCE_EXPIRATION_IN_MILLISECONDS)
                .setTransitionTypes(Geofence.GEOFENCE_TRANSITION_ENTER | Geofence.GEOFENCE_TRANSITION_EXIT)
                .build();

        try
        {
            LocationServices.GeofencingApi.addGeofences(googleApiClient, getGeofencingRequest(geofence), getGeofencePendingIntent())
                    .setResultCallback(new ResultCallback<Status>()
                    {
                        @Override
                        public void onResult(@NonNull Status status)
                        {
                            if (status.isSuccess()) {
                                //db.addGeoLocation(lat,lng,radius,expire,key);
                            } else {
                                //String errorMessage = GeofenceTrasitionService.getErrorString(MainActivity.this, status.getStatusCode());
                                //Log.e(TAG, errorMessage);
                            }
                        }
                    });

        } catch (SecurityException securityException) {
            logSecurityException(securityException);
        }
    }

    private GeofencingRequest getGeofencingRequest(Geofence geofence) {
        GeofencingRequest.Builder builder = new GeofencingRequest.Builder();
        builder.setInitialTrigger(GeofencingRequest.INITIAL_TRIGGER_ENTER);
        builder.addGeofence(geofence);
        return builder.build();
    }

    private PendingIntent getGeofencePendingIntent() {
        if (mGeofencePendingIntent != null) {
            return mGeofencePendingIntent;
        }
        Intent intent = new Intent(getActivity(), GeofenceTrasitionService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            return PendingIntent.getService(getActivity(), 0, intent, PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);
        }else {
            return PendingIntent.getService(getActivity(), 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
        }
    }

    @Override
    public boolean onMarkerClick(Marker marker) {
        //System.out.println("Title : " + marker.getTitle());
        return false;
    }

     @Override
    public void onInfoWindowClick(Marker marker) {
        final String key = marker.getTitle().split(":")[1];

         try{
             //googleMap.clear();
             removeGeofenceMark(key);
         }
         catch(Exception e){
         }
    }

    public void removeGeofenceMark(final String key)
    {
        if (!googleApiClient.isConnected()) {
            Toast.makeText(getActivity(), "GeoFence Not connected!", Toast.LENGTH_SHORT).show();
            return;
        }

        try
        {
            List<String> idList = new ArrayList<>();
            idList.add(key);
            LocationServices.GeofencingApi.removeGeofences(googleApiClient, idList)
                    .setResultCallback(new ResultCallback<Status>()
                    {
                        @Override
                        public void onResult(@NonNull Status status)
                        {
                            if (status.isSuccess()) {
                                Toast.makeText(getActivity(), "GeoFence removed successfully !", Toast.LENGTH_SHORT).show();

                                googleMap.clear();
                            } else {
                                // Get the status code for the error and log it using a user-friendly message.
                                String errorMessage = GeofenceTrasitionService.getErrorString(getActivity(), status.getStatusCode());
                                Log.e(TAG, errorMessage);
                                Toast.makeText(getActivity(), "Error removing GeoFence !" + errorMessage, Toast.LENGTH_SHORT).show();
                            }
                        }
                    });
        } catch (SecurityException securityException) {
            logSecurityException(securityException);
        }
    }

    private void logSecurityException(SecurityException securityException) {
        Log.e(TAG, "Invalid location permission. " +
                "You need to use ACCESS_FINE_LOCATION with geofences", securityException);
    }

}

File: android_samples_kotlin/04 GoogleMapDemo/app/src/main/java/com/msnegi/googlemapdemo/HomeFragment.kt
package com.msnegi.googlemapdemo

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.msnegi.googlemapdemo.databinding.FragmentHomeBinding

import android.Manifest
import android.app.Activity.RESULT_CANCELED
import android.app.Activity.RESULT_OK
import android.app.AlertDialog
import android.content.Context
import android.content.Context.LOCATION_SERVICE
import android.content.DialogInterface
import android.content.Intent
import android.content.IntentSender.SendIntentException
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import android.util.Log
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.common.api.ResultCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.LocationSettingsRequest
import com.google.android.gms.location.LocationSettingsResult
import com.google.android.gms.location.LocationSettingsStatusCodes
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.model.Marker
import com.msnegi.locationdemo.LatLongReceived
import com.msnegi.locationdemo.LocManager

class HomeFragment : Fragment(), OnMapReadyCallback, LatLongReceived {

    private val permission = arrayOf(
        Manifest.permission.ACCESS_FINE_LOCATION,
    )

    companion object {
        private const val PERMISSIONS_REQUEST_CODE = 101
        private const val TAG = "HomeFragment"
    }

    lateinit var googleApiClient: GoogleApiClient
    //lateinit var mapView: MapView
    var googleMap: GoogleMap? = null
    //lateinit  var locationMarker: Marker
    var zoomLevel: Float = 15f
    var locationManager: LocManager? = null

    private var _binding: FragmentHomeBinding? = null

    // This property is only valid between onCreateView and onDestroyView.
    private val binding get() = _binding!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        googleApiClient = GoogleApiClient.Builder(requireActivity())
                                         .addApi(LocationServices.API).build();
        googleApiClient.connect();

        if (ContextCompat.checkSelfPermission(
                activity,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
            || ContextCompat.checkSelfPermission(
                activity,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                activity,
                permission,
                PERMISSIONS_REQUEST_CODE
            )
        } else {
            //-- permissions are already given
            checkGPSSettings()
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentHomeBinding.inflate(inflater, container, false)
        _binding?.mapView!!.onCreate(savedInstanceState)
        _binding?.mapView!!.getMapAsync(this)
        return binding.root
    }

    override fun onMapReady(mMap: GoogleMap) {
        googleMap = mMap;
        _binding?.mapView!!.onResume();

        //if (googleMap != null) {
        //     setupMap(28.5022851, 77.0857269)
        //}
    }

    private fun setupMap(mLat: Double, mLong: Double) {
        val mLoc = LatLng(mLat, mLong)

        googleMap!!.clear()
        val options = MarkerOptions()
        options.position(mLoc)
        options.icon(BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_GREEN))
        googleMap!!.addMarker(options)
        googleMap!!.getUiSettings().setZoomControlsEnabled(true)
        googleMap!!.moveCamera(CameraUpdateFactory.newLatLng(mLoc))
        googleMap!!.animateCamera(CameraUpdateFactory.newLatLngZoom(LatLng(mLat, mLong), 5f))

        val cameraPosition = CameraPosition.Builder()
            .target(LatLng(mLat, mLong))
            .zoom(zoomLevel)
            .build()
        googleMap!!.animateCamera(CameraUpdateFactory.newCameraPosition(cameraPosition))
        _binding?.mapView!!.onResume()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    //---- locations code

    fun checkGPSSettings() {
        if (isGPSEnabled(getContext())) {
            Toast.makeText(activity, "GPS enabled", Toast.LENGTH_SHORT).show()

            locationManager = LocManager.getInstance()
            locationManager!!.connectGoogleClient(requireContext(), this)
        } else {
            Toast.makeText(activity, "GPS Disabled", Toast.LENGTH_SHORT).show()
            displayLocationSettingsRequest(requireContext())
        }
    }

    override fun onLocationReceived(loc: Location?) {
        println("First Lat : " + loc?.getLatitude() + " Long : " + loc?.getLongitude())
        //_binding.txtLatLong!!.setText("Lat : " + loc?.getLatitude() + " Long : " + loc?.getLongitude())
        if (googleMap != null) {
            setupMap(loc?.getLatitude() ?: 0.0, loc?.getLongitude() ?: 0.0 );
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        when (requestCode) {
            PERMISSIONS_REQUEST_CODE -> {
                var permission_denied = false
                var i = 0
                while (i < permissions.size) {
                    if (grantResults[i] == PackageManager.PERMISSION_DENIED) {
                        permission_denied = true
                        if (ActivityCompat.shouldShowRequestPermissionRationale(
                                activity,
                                permissions[i]!!
                            )
                        ) {
                            showAlert()
                        }
                    }
                    i++
                }
                if (!permission_denied) {
                    //-- move on to track location
                    checkGPSSettings()
                }
            }
        }
    }

    private fun showAlert() {
        val alertDialog = AlertDialog.Builder(activity).create()
        alertDialog.setTitle("Alert")
        alertDialog.setMessage("App needs all required permissions to run functionalities !!!")
        alertDialog.setButton(
            AlertDialog.BUTTON_NEGATIVE, "Do not Allow",
            object : DialogInterface.OnClickListener {
                override fun onClick(dialog: DialogInterface, which: Int) {
                    Toast.makeText(
                        activity,
                        "App required all permission to work properly",
                        Toast.LENGTH_SHORT
                    ).show()
                    dialog.dismiss()
                    //-- no action required, move on to application
                }
            })
        alertDialog.setButton(
            AlertDialog.BUTTON_POSITIVE,
            "Allow",
            object : DialogInterface.OnClickListener {
                override fun onClick(dialog: DialogInterface, which: Int) {
                    dialog.dismiss()
                    ActivityCompat.requestPermissions(
                        activity,
                        permission,
                        PERMISSIONS_REQUEST_CODE
                    )
                }
            })
        alertDialog.show()
    }

    fun isGPSEnabled(context: Context?): Boolean {
        val manager = context?.getSystemService(LOCATION_SERVICE) as LocationManager
        val isGPSEnabled = manager.isProviderEnabled(LocationManager.GPS_PROVIDER)
        return isGPSEnabled
    }

    private fun displayLocationSettingsRequest(context: Context) {
        val googleApiClient = GoogleApiClient.Builder(context)
            .addApi(LocationServices.API).build()
        googleApiClient.connect()

        val locationRequest = LocationRequest.create()
        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)
        locationRequest.setInterval(10000)
        locationRequest.setFastestInterval((10000 / 2).toLong())

        val builder = LocationSettingsRequest.Builder().addLocationRequest(locationRequest)
        builder.setAlwaysShow(true)

        val result =
            LocationServices.SettingsApi.checkLocationSettings(googleApiClient, builder.build())
        result.setResultCallback(object : ResultCallback<LocationSettingsResult?> {
            override fun onResult(result: LocationSettingsResult) {
                val status = result.getStatus()
                when (status.getStatusCode()) {
                    LocationSettingsStatusCodes.SUCCESS -> Log.i(
                        TAG,
                        "All location settings are satisfied."
                    )

                    LocationSettingsStatusCodes.RESOLUTION_REQUIRED -> try {
                        // Show the dialog by calling startResolutionForResult(), and check the result
                        // in onActivityResult().
                        status.startResolutionForResult(requireActivity(), 8576)
                    } catch (e: SendIntentException) {
                        Log.i(TAG, "PendingIntent unable to execute request.")
                    }

                    LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE -> Log.i(
                        TAG,
                        "Location settings are inadequate, and cannot be fixed here. Dialog not created."
                    )
                }
            }
        })
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 8576 && resultCode == RESULT_OK) {
            Toast.makeText(activity, "gps enabled", Toast.LENGTH_SHORT).show()

            locationManager = LocManager.getInstance()
            locationManager!!.connectGoogleClient(requireContext(), this)
        } else if (requestCode == 8576 && resultCode == RESULT_CANCELED) {
            Toast.makeText(activity, "gps disabled", Toast.LENGTH_SHORT).show()
            //finish();
        }
    }

    //-- regularly receive the lat-longs (implementing in next build)

}